File: cpp/leetcode/381.cpp
class RandomizedCollection {
public:
    RandomizedCollection() { }
    
    bool insert(int val) {
        m[val].push_back(nums.size());
        nums.push_back(make_pair(val, m[val].size() - 1));
        return m[val].size() == 1;
    }
    
    bool remove(int val) {
        auto result = m.find(val) != m.end();
        if(result) {
            int idx = m[val].back();
            m[val].pop_back();
            int last = nums.back();
            m[last.first][last.second] = idx;
            nums[idx] = last;
            if(m[val].empty()) m.erase(val);
            nums.pop_back();
        }
        return result;
    }
    
    int getRandom() {
        return nums[rand() % nums.size()].first;
    }
private:
    vector<pair<int, int>> nums;
    unordered_map<int, vector<int>> m;
};



class RandomizedCollection {
public:
    RandomizedCollection() {}
    
    bool insert(int val) {
        nums.push_back(val);
        m[val].push_back(nums.size() - 1);
        return m[val].size() == 1;
    }
    
    bool remove(int val) {
        if(m.find(val) != m.end()){
            int idx = m[val].back();
            m[val].pop_back();
            if(m[val].empty()) m.erase(val);
            if(nums.size() - 1 != idx){
                int last = nums.back();
                nums[idx] = last;
                m[last].pop_back();
                m[last].insert(m[last].begin(), idx);
            }
            nums.pop_back();
            return true;
        }
        return false;
    }
    
    int getRandom() {
        return nums[rand() % nums.size()];
    }
private:
    vector<int> nums;
    unordered_map<int, vector<int>> m;
};

File: python/leetcode/381.py
class RandomizedCollection(object):
    def __init__(self):
        self.arr = []
        self.map = collections.defaultdict(set)

    def insert(self, val):
        self.map[val].add(len(self.arr))  # add val: index to dict
        self.arr.append(val)            # append val to list
        return len(self.map[val]) == 1

    def remove(self, val):
        if val not in self.map: return False
        idx, newVal = self.map[val].pop(), self.arr[-1] 
        # d[val].pop as set.pop, last element from d[val]
        # del d[val]'s index
        if len(self.map[val]) == 0: del self.map[val]
        self.arr[idx] = newVal  # overlap val in i
        if newVal in self.map:
            self.map[newVal].add(i)
            # s.discard(x) removes x from set s if present
            # remove the last idx, which is swap for val
            self.map[newVal].discard(len(self.arr)-1) 
        self.arr.pop() # remove val completely
        return True

    def getRandom(self):
        return random.choice(self.arr)  # same as self.arr[randint(0, n-1)]